artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireCoreConverter.java

```diff
@@ -0,0 +1,572 @@
+      if (groupId != null) {
+         coreMessage.putStringProperty(org.apache.activemq.artemis.api.core.Message.HDR_GROUP_ID,
+               new SimpleString(groupId));
+         coreMessage.putStringProperty(JMS_GROUPID, new SimpleString(groupId));
+      }
+      if (openwireMessage.containsProperty(JMS_GROUPSEQ)) {
+         coreMessage.putIntProperty(JMS_GROUPSEQ, openwireMessage.getIntProperty(JMS_GROUPSEQ));
+         coreMessage.putIntProperty(JMS_XGROUPSEQ, openwireMessage.getIntProperty(JMS_GROUPSEQ));
+      }
+      final SimpleString lastValueProperty = openwireMessage.getLastValueProperty();
+      if (lastValueProperty != null) {
+         coreMessage.setLastValueProperty(lastValueProperty);
+      }
+      coreMessage.putObjectProperty(AMQP_CONTETTYPE, openwireMessage.getObjectProperty(JMS_CONTENT_TYPE));
+      coreMessage.putObjectProperty(AMQP_REPLYTOGROUPID, openwireMessage.getObjectProperty(AMQP_REPLYTOGROUPID));
+      coreMessage.setReplyTo(openwireMessage.getReplyTo());
+      coreMessage.setUserID(openwireMessage.getUserID());
+      coreMessage.setDurable(openwireMessage.isDurable());
+      coreMessage.setAddress(openwireMessage.getAddress());
+      coreMessage.putObjectProperty(org.apache.activemq.artemis.api.core.Message.HDR_GROUP_ID,
+            (SimpleString) openwireMessage
+                  .getObjectProperty(org.apache.activemq.artemis.api.core.Message.HDR_GROUP_ID));
+      copyAllProperties(openwireMessage, coreMessage);
+      return coreMessage;
+   }
+
+   public static byte toCoreType(byte amqType) {
+      switch (amqType) {
+         case CommandTypes.ACTIVEMQ_BLOB_MESSAGE:
+            throw new IllegalStateException("We don't support BLOB type yet!");
+         case CommandTypes.ACTIVEMQ_BYTES_MESSAGE:
+            return org.apache.activemq.artemis.api.core.Message.BYTES_TYPE;
+         case CommandTypes.ACTIVEMQ_MAP_MESSAGE:
+            return org.apache.activemq.artemis.api.core.Message.MAP_TYPE;
+         case CommandTypes.ACTIVEMQ_OBJECT_MESSAGE:
+            return org.apache.activemq.artemis.api.core.Message.OBJECT_TYPE;
+         case CommandTypes.ACTIVEMQ_STREAM_MESSAGE:
+            return org.apache.activemq.artemis.api.core.Message.STREAM_TYPE;
+         case CommandTypes.ACTIVEMQ_TEXT_MESSAGE:
+            return org.apache.activemq.artemis.api.core.Message.TEXT_TYPE;
+         case CommandTypes.ACTIVEMQ_MESSAGE:
+            return org.apache.activemq.artemis.api.core.Message.DEFAULT_TYPE;
+         default:
+            throw new IllegalStateException("Unknown ActiveMQ Artemis message type: " + amqType);
+      }
+   }
+
+   public static void loadMapIntoProperties(TypedProperties props, Map<String, Object> map) {
+      for (Entry<String, Object> entry : map.entrySet()) {
+         SimpleString key = new SimpleString(entry.getKey());
+         Object value = entry.getValue();
+         if (value instanceof UTF8Buffer) {
+            value = ((UTF8Buffer) value).toString();
```

Comment:
This cast is redundant: it was in the original code too, but it could dropped now